Automate releases with `release.yaml`

cisaacstern opened the pull request:
Currently:

- latest release according to pypi is `0.7.1`
- latest tag on GitHub is `v0.7.0`
- latest release according to GitHub releases page is `0.6.1`

🙃 

This PR adds the same `release.yaml` as used in `pangeo-forge-recipes` to automate this process.

@yuvipanda, if this looks good to you, I can add a pypi token as a repo secret and cut a `0.8.0` release.

yuvipanda commented:
@cisaacstern you can use pypi's trusted publisher instead, is easier! https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers/

Feel free to self-merge once you set that up! Here's an example PR in another project: https://github.com/jupyterhub/jupyter-remote-desktop-proxy/pull/46

cisaacstern commented:
@yuvipanda thanks, didn't know about trusted publishing, it's great! Added that.

Pinging you for a quick re-review because I also changed a few other things to enable dynamic versioning... does that stuff look okay to you?

yuvipanda commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,33 @@
+
+on:
+  release:
+    types: [created]
```
I think this should get triggered on tag creation instead: https://github.com/jupyterhub/jupyter-remote-desktop-proxy/pull/46/files#diff-e426ed45842837026e10e66af23d9c7077e89eacbe6958ce7cb991130ad05adaL22

Releases are github specific, tags are universal to git!

cisaacstern replied:
If we trigger on tag creation, do we expose ourselves to the risk that the latest release displayed on the GitHub landing page is out of sync with the latest release on pypi? (Which I think is our current situation on this repo.)

On -recipes, we release by clicking "make a new release" on GitHub, and then create the new tag as part of that process.

If this is triggered instead on tag, is the thought that the tags are still created on the GitHub browser UI? And then how do we keep GitHub releases in sync with tags... another step in this workflow that adds the release after the tag is made, following successful publish to pypi?

yuvipanda replied:
I actually think nobody in the Jupyter project actually use releases, just tags - see https://github.com/jupyterhub/jupyterhub for example. 
<img width="205" alt="image" src="https://github.com/pangeo-forge/pangeo-forge-runner/assets/30430/6a5b6200-0872-4cba-a689-b6e9433670c4">

This made me realize I've never actually used "Releases" as a feature outside of tags. So I think my recommendation didn't actually have anything to do with github vs git, but just that I have no idea what the point of the "Releases" feature is! I know you can upload binary builds if you have 'em, but we don't. So I guess inherent in the suggestion to use tags is that - I don't know what the use case for GitHub Releases are, and I'm not sure why we would use that feature at all in any way or form! So there would actually be *fewer* steps with tags, as we would just not use Releases at all :D

cisaacstern replied:
Haha! Okay this makes sense, tags it is 💫 😄

yuvipanda reviewed (changes requested):
One comment about when to trigger, but otherwise if this works go for it! Thanks for taking care of it, @cisaacstern!

cisaacstern commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,33 @@
+name: Release
+
+on:
+  release:
+    types: [created]
```
```suggestion
  push:
    tags: ["**"]
```

yuvipanda reviewed (approved):
Thanks for indulging me, @cisaacstern! Feel free to merge whenever you feel like
